File: 0189-rotate-array/0189-rotate-array.py
class Solution:
    def rotate(self, nums: List[int], k: int) -> None:
        """
        Do not return anything, modify nums in-place instead.
        """
        # Honestly, I would like to use a ring buffer here instead of raw vector
        
        # Normalize indexes
        n = len(nums)
        k = k % n 
        
        # Triple reversal algorithm.
        nums[:] = nums[-k:] + nums[:-k]

File: 0189-rotate-array/0189-rotate-array.cpp
#include <vector>
#include <algorithm> 
#include <iostream>


class Solution {

public:
    void rotate(vector<int>& nums, int k) {
        std::ios_base::sync_with_stdio(false);
        std::cin.tie(NULL);

        if (nums.empty()) {
            return;
        }

        // Normalize indexes
        int n = nums.size();
        k = k % n;

        // Idle if k = 0
        if (k == 0) {
            return;
        }

        // Find border between region A and region B for the Triple reversal algorithm.
        auto new_start_iterator = nums.begin() + (n - k);
        std::rotate(nums.begin(), new_start_iterator, nums.end());
    }
};



